### ISSUE TITLE
Keep polling the car during "charging"

### ISSUE THREAD

**Issue by dilorenzo1987** [alandtse/tesla#426]
Currently, it may happen that the car finishes charging but goes to sleep before the integration can poll the vehicle again. It's then disconnected but the Charging status will stay until the car awakes again for whatever reason.

I propose to keep polling the car and ignore the disconnected state as long as the last charging status is "Charging".

**Comment by carleeno** [comment]
Unfortunately it can't poll if it's asleep without waking it. But also it shouldn't go to sleep if it is still being polled regularly (as it would be while charging). What is your poll rate set to? If it's too long then it may have time to sleep between regular polls.

I recommend going with a shorter poll interval so that it won't go to sleep before ha updates from charging to idle, at which point it'll automatically slow the polling to let the car sleep (if you have it on the 'normal' profile).

**Comment by dilorenzo1987** [comment]
Hi Carleeno

That's actually what I think would be nice, to wake up the car in such cases.
This would fix that issue without the need to increase the polling interval unnecessarily.
Further, it will only happen if this "rare" case happens.

Regards

**Comment by InTheDaylight14** [comment]
In a similar vein, is it possible to poll more frequently while driving? Maybe every minute while driving and then the setup frequency while not in drive?

**Comment by carleeno** [comment]
@InTheDaylight14 yes it does already. By default it polls every 60 seconds while driving, 300 seconds when parked and not trying to sleep, and 660 seconds when trying to sleep.

I think the answer might be to decrease the charging polling interval just like it does for driving, so that we get an update to understand if it's finished charging before it goes to sleep, and then slow down the polling rate back to normal.

I find it unusual that the car would go to sleep within < 5 minutes of finishing charging, given the default polling rate, but I have my polling interval set to 15 seconds so I never noticed this as it will finish charging, and then it'll go from 15 seconds to 660 seconds to let it sleep.

I have a WIP pr which updates the online/sleep logic and I'll see if I can find a way to address this post-charging issue without waking the car just to poll it. Otherwise I think it would be risky in causing the integration to start keeping cars awake when not desired.

**Comment by dilorenzo1987** [comment]
> 300 seconds when parked and not trying to sleep, and 660 seconds when trying to sleep.

How is that determined (whether it's trying to sleep or not), I always thought the integration will switch to 660 immediately after driving to let the car sleep.?

> I think the answer might be to decrease the charging polling interval just like it does for driving, so that we get an update to understand if it's finished charging before it goes to sleep, and then slow down the polling rate back to normal.

Sounds like a solution :)

**Comment by InTheDaylight14** [comment]
> By default it polls every 60 seconds while driving, 300 seconds when parked and not trying to sleep, and 660 seconds when trying to sleep.
> 

Oh cool! Would you mind pointing me to where that happens in the code? Just curious to read through it.

**Comment by carleeno** [comment]
@InTheDaylight14 all of these time defaults are defined in teslajsonpy/const.py, and used in controller.py

**Comment by alandtse** [comment]
It should also be doing 60 seconds during charging or when climate is on. Please reference the wiki.

**Comment by carleeno** [comment]
> It should also be doing 60 seconds during charging or when climate is on. Please reference the wiki.

@alandtse maybe the wiki is out of date? the code doesn't do this, I double checked. it only polls vehicle data at 60 second intervals when the vehicle is in gear.

**Comment by alandtse** [comment]
I may be misremembering but the Wiki should be considered the spec we adhere to. If we don't match it, then it's a bug. We'll need to see about reminding PRs to include wiki updates. 

Looking at it again, https://github.com/alandtse/tesla/wiki/Polling-policy, it looks like it stuck to polling_interval while charging. I am ok with a PR taking it down to 60 seconds. The car isn't sleeping so the energy cost should be minimal. The other alternative is to schedule an update at the last known estimated completion time since we have that now.

**Comment by evolucja** [comment]
Do we have any tests made with all those different pooling settings?  I mean that we have 3 different policies: normal/connected/always and additionally we can set different pooling intervals. I am really considering how all of them actually impact the battery life during standby/charging/driving. 
Maybe we shouldn't be so restrict...

**Comment by dilorenzo1987** [comment]
Today my logbook got spammed probably due to this behavior.
![2022-12-24 12_00_38-Photo - Google Photos](https://user-images.githubusercontent.com/12641114/209432926-741c84c3-b589-4073-bd1c-5d785d37ba39.png)

The charging session was short (got Tesla app push notifications) and already finished by that time but the car stayed on "Charging".

Please let me know if I can help with any log files.

**Comment by InTheDaylight14** [comment]
> Today my logbook got spammed probably due to this behavior. ![2022-12-24 12_00_38-Photo - Google Photos](https://user-images.githubusercontent.com/12641114/209432926-741c84c3-b589-4073-bd1c-5d785d37ba39.png)
> 
> The charging session was short (got Tesla app push notifications) and already finished by that time but the car stayed on "Charging".
> 
> Please let me know if I can help with any log files.

Polling every minute might fix that issue but it's really driven by poor logic when I first wrote the sensor. There is a PR https://github.com/alandtse/tesla/pull/419 ready to fix the time charge complete updating every minute like in these logs.

**Comment by evolucja** [comment]
@mikkelzzz3 thanks for the approach, I was also trying to change the update interval
